### DISCUSSION TITLE
How do I addColumn where a default value is required?

### DISCUSSION THREAD

**Discussion by sharathpanily** [simolus3/drift#2312] [Q&A] [answered]
I'm adding a column to an existing table using `await m.addColumn(table1, table1.column2);` and I'm getting the following error:

```bash
DatabaseException(error code 1: Cannot add a NOT NULL column with default value NULL) sql 'ALTER TABLE "table1" ADD COLUMN "column2" INTEGER NOT NULL REFERENCES table2 (identifier);' args []
```

I assume this is because the `identifier` column of table2 is non-nullable and this newly added column2 references that column - how do I fix this error - or in other words, how do I provide a default value? Code for `identifier` column is referenced below.

```dart
  IntColumn get identifier => integer().autoIncrement()();
```
And for reference I'm using `drift: ^2.5.0` and flutter 3.7.0

Thanks for your help!

**Comment by simolus3** [comment, marked as answer]
The error message is about `table1.column2`, which was declared to be non-nullable. That it references another column is not the problem.

There is no direct way to work around that error. If you have existing rows in `table1` and then add your `column2`, what values should the existing rows have in that column? It can't be `NULL`. You can make the column `.nullable()` if that's what you want, but otherwise you need to come up with a reasonable default.

If you have a default here, you can use drift's `alterTable` utility to add the new column:

```dart
await m.alterTable(
  TableMigration(
    table1,
    newColumns: [table1.column2],
    columnTransformer: {
      table1.column2: const Constant(1), // if you wanted to have 1 as a default value
    },
  )
);
```

**Reply by sharathpanily** [reply]
Thanks for your help! Before you responded, I tried the following code in specifying `table1.column2` and it seemed to work.

```dart
  IntColumn get column2 => integer().withDefault(const Constant(-1)).references(table2, #identifier)();
```

Is this a reasonable or acceptable solution as well?

**Reply by simolus3** [reply]
If you want the `-1` to apply for new values as well, then yes! That's the best solution in that case. If you only want the default to apply to existing rows at the time you add the new column, you'd have to use a table migration.

**Reply by sharathpanily** [reply]
I want to make sure I fully understand each approach - using `withDefault` while specifying `table1.column2` only specifies the default value if no value is provided - correct? It does not mean that the value of `table1.column2` cannot be changed during/after row creation?

**Reply by simolus3** [reply]
Exactly! `withDefault` only applies to new rows inserted into the table if no other value of `column2` has been explicitly set. You can still update values for the column with `update` later.
